llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,322 @@
+      // We need to invoke the default constructor on all the elements of the
+      // page.
+      for (std::size_t I = 0; I < PAGE_SIZE; ++I)
+        new (NewPagePtr + I) T();
+
+      PagePtr = reinterpret_cast<uintptr_t>(NewPagePtr);
+    }
+    // Dereference the element in the page.
+    return *((Index % PAGE_SIZE) + reinterpret_cast<T *>(PagePtr));
+  }
+
+  // Return the capacity of the vector. I.e. the maximum size it can be expanded
+  // to with the resize method without allocating more pages.
+  [[nodiscard]] std::size_t capacity() const {
+    return PageToDataIdx.size() * PAGE_SIZE;
+  }
+
+  // Return the size of the vector. I.e. the maximum index that can be
+  // accessed, i.e. the maximum value which was used as argument of the
+  // resize method.
+  [[nodiscard]] std::size_t size() const { return Size; }
+
+  // Expands the vector to the given NewSize number of elements.
+  // If the vector was smaller, allocates new pages as needed.
+  // It should be called only with NewSize >= Size.
+  void resize(std::size_t NewSize) {
+    // Handle shrink case: delete the pages and update the size.
+    if (NewSize < Size) {
+      std::size_t NewLastPage = (NewSize - 1) / PAGE_SIZE;
+      for (std::size_t I = NewLastPage + 1; I < PageToDataIdx.size(); ++I) {
+        uintptr_t PagePtr = PageToDataIdx[I];
+        if (PagePtr == InvalidPage)
+          continue;
+        T *Page = reinterpret_cast<T *>(PagePtr);
+        // We need to invoke the destructor on all the elements of the page.
+        for (std::size_t J = 0; J < PAGE_SIZE; ++J)
+          Page[J].~T();
+        Allocator.getPointer()->Deallocate(Page);
+      }
+      // Delete the extra ones in the new last page.
+      uintptr_t PagePtr = PageToDataIdx[NewLastPage];
+      if (PagePtr != InvalidPage) {
+        T *Page = reinterpret_cast<T *>(PagePtr);
+        // If the new size and the old size are on the same page, we need to
+        // delete only the elements between the new size and the old size.
+        // Otherwise we need to delete all the remaining elements in the page.
+        std::size_t OldPage = (Size - 1) / PAGE_SIZE;
+        std::size_t NewPage = (NewSize - 1) / PAGE_SIZE;
+        std::size_t LastPageElements =
+            OldPage == NewPage ? Size % PAGE_SIZE : PAGE_SIZE;
+        for (std::size_t J = NewSize % PAGE_SIZE; J < LastPageElements; ++J)
+          Page[J].~T();
+      }
+      PageToDataIdx.resize(NewLastPage + 1);
+    }
+    Size = NewSize;
+    // If the capacity is enough, just update the size and continue
+    // with the currently allocated pages.
+    if (Size <= capacity())
+      return;
+    // The number of pages to allocate. The Remainder is calculated
+    // for the case in which the NewSize is not a multiple of PAGE_SIZE.
+    // In that case we need one more page.
+    std::size_t Pages = Size / PAGE_SIZE;
+    std::size_t Remainder = Size % PAGE_SIZE;
+    if (Remainder != 0)
+      Pages += 1;
```

Comment:
You can reuse the `NewLastPage` / `NewNumPages` calculation from the start of this function rather than doing it this way.

Reply:
I've added some suggested edits to demonstrate.